### PR TITLE
Gpx

### PR DESCRIPTION
**Pull request by vcoindet**
#114 :

GPX_Provider created : it parses GPX files (gets Waypoints and Tracks points) and creates Three geometies that can be added to the scene (points and lines). 
It can be called with the 'loadGPX' function wich is in ApiGlobe.js.

One bug is still to be fixed on Windows : 'linewidth' attribut doesn't have any effect on Windows (see #155 ).

### CODE REVIEW

**Comment by nosy-b** [comment]
Super!
We'll check for LineWidth and other style representation for features soon. Do you have a screenshot of a GPX visualization?

**Comment by vcoindet** [comment]
gpx

**Comment by qdnguyen** [comment]
how do you activate this layer? i did not see on GUI?

**Comment by vcoindet** [comment]
gpx2

**Comment by vcoindet** [comment]
You have to call it in index.html

```
viewerDiv.addEventListener('globe-loaded',function(){
                    console.log("Globe Loaded");
                    itowns.viewer.loadGPX("URL_of_your_file");
                } ,
 false);
```

**Comment by qdnguyen** [comment]
Can you publish the data in https://github.com/iTowns/iTowns2-sample-data ? juste one file to test.

otherwise, why the data is loaded in addEventListener? we can not have a provider like wfs?

**Comment by vcoindet** [comment]
You don't have to put in an addEventListener for this one.
The event 'globe-loaded' is here for some other functions, like 'setCenter' where you have to wait for the globe to be full loaded or you will have some accuracy errors. Here it's not necessary.

**Comment by qdnguyen** [comment]
But i dont understand how you deal with many files? or just only one file we need in itowns?

**Comment by vcoindet** [comment]
For now, you can load only one file at the time, or call the function several times.

**Comment by vcoindet** [comment]
I just found a new problem in my code : 
I'm using the tag "ele" in gpx files to get the elevation, but sometime this tag doesn't exist, so my algorithm doesn't work.

This will be fixed when 'Globe.getCurrentAltitude(longitude, latitude)' is done. (see issue #119)

File: src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js [hunk 1]
```diff
@@ -639,6 +641,18 @@ define('Core/Commander/Interfaces/ApiInterface/ApiGlobe', [
     };
 
 
+    ApiGlobe.prototype.loadGPX = function(url) {
+        //providerGPX should be instanced elsewhere
```
Comment (peppsac) [inline src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js]: > -        //providerGPX should be instanced elsewhere 

Indeed. All other providers are instanced in `createSceneGlobe`

File: src/Core/Commander/Providers/GPX_Provider.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                // ------------------------------------
+                // Getting the waypoint points
+                // ------------------------------------
+                var wpt = result.getElementsByTagName("wpt");
+                var geometry_p = new THREE.Geometry();
+
+                for (var i = 0; i < wpt.length; i++)
+                    geometry_p.vertices.push(this.ellipsoid.cartographicToCartesian(new CoordCarto().setFromDegreeGeo(Number(wpt[i].attributes.lon.nodeValue),Number(wpt[i].attributes.lat.nodeValue),Number(wpt[i].getElementsByTagName("ele")[0].childNodes[0].nodeValue))));
+
+                var material_p = new THREE.PointsMaterial( {color: 0x00ff00, size: 50, sizeAttenuation : true} );
```
Comment (peppsac) [inline src/Core/Commander/Providers/GPX_Provider.js]: You probably don't need to create 1 PointMaterial per point. 
Suggestion: move Materials creation before the `for (var i = 0; i < wpt.length; i++)` loop

Comment (peppsac) [inline src/Core/Commander/Providers/GPX_Provider.js]: Oops nevermind I misread.
New coment : please use `{ }` for each block :)

File: src/Core/Commander/Providers/GPX_Provider.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+                var point = new THREE.Points( geometry_p, material_p );
+
+                // ------------------------------------
+                //Getting the track points
+                // ------------------------------------
+                var trkpt = result.getElementsByTagName("trkpt");
+                var geometry = new THREE.Geometry();
+
+                for (var j = 0; j < trkpt.length; j++)
```
Comment (peppsac) [inline src/Core/Commander/Providers/GPX_Provider.js]: Add braces `{ }`

File: src/Core/Commander/Providers/GPX_Provider.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                var line = new THREE.Line(geometry, material);
+                line.add(point);
+
+                return line;
+            }.bind(this));
+
+        };
+
+        return GPX_Provider;
+
```
Comment (peppsac) [inline src/Core/Commander/Providers/GPX_Provider.js]: Lots of unneeded blank lines in this file. Can you remove them please?

**Comment by peppsac** [comment]
What about making a real Provider (one that can execute commands)?

Alternatively : split the parseGpx function to a GpxUtil.js file and do:

``` js
xmlDriver.read(url).then(function(xml) {
    this.scene.getMap().gpxTracks.children[0].add(GPXUtils.GpxToGeometry(xml));
}.bind(this));
```

in Globe.js

File: src/Core/Commander/Providers/GPX_Provider.js (hunk 4 above)
Comment (qdnguyen) [inline src/Core/Commander/Providers/GPX_Provider.js]: i think you should use group = new Object3D and return group instead of line.

**Comment by qdnguyen** [comment]
finaly, we can see line width styled with texture under Angle. 
autoroute1

**Comment by qdnguyen** [comment]
autoroute1

File: src/Core/Commander/Providers/GPX_Provider.js (hunk 4 above)
Comment (qdnguyen) [inline src/Core/Commander/Providers/GPX_Provider.js]: done in my code!

**Comment by qdnguyen** [comment]
add ItownsPoint module, i will use this for potree also!
bb

**Comment by gchoqueux** [comment]
GPX is broken, for the moment

**Comment by qdnguyen** [comment]
because my last commit is changed?

**Comment by gchoqueux** [comment]
r?

**Comment by peppsac** [comment]
- Indentation is incorrect in several places
- Missing explanations about _why_ you need to reimplement Points/Lines
- No answer to my comment about the GpxProvider: https://github.com/iTowns/itowns2/pull/158#issuecomment-234537490

**Comment by qdnguyen** [comment]
> Indentation is incorrect in several places

Can you capture a photo please? at some zoom level, line is disappeared but i dont know why? 

> Missing explanations about why you need to reimplement Points/Lines

For line, you can see some explainations here #155. Another reason I styled line is that we can add street name as what they did in google street view :

sans titre

For points, we dont need to create a specifique shader to show texture but i did it because i will extend this shader for potree also. 

> No answer to my comment about the GpxProvider: #158 (comment)

I leave @vcoindet to reply you this question because it relate to his work but he is in vacation now.

**Comment by gchoqueux** [comment]
I agree this comment : https://github.com/iTowns/itowns2/pull/158#issuecomment-234537490
I'm working on it.

**Comment by gchoqueux** [comment]
> I agree this comment : https://github.com/iTowns/itowns2/pull/158#issuecomment-234537490

Update branch : Recast Gpx provider

**Comment by peppsac** [comment]
I pushed a commit that take advantages of modules.
The GpxUtils module now exports a single function `loadGpx` and all others are private.

Another question : GpxUtils.js should probably move elsewhere (`src/utils/` ?) but this can wait until we have at least two _utils_ modules.

**Pull request merged**